src/bounded.rs

```diff
@@ -225,6 +225,54 @@ impl<T> Bounded<T> {
         }
     }
 
+    pub fn get_head(&mut self) -> Result<T, GetError> {
+        let head = self.head.load(Ordering::SeqCst);
+        let tail = self.tail.load(Ordering::SeqCst);
+
+        // If the tail equals the head, that means the queue is empty.
+        if (tail & !self.mark_bit) == head {
+            // Check if the queue is closed.
+            if tail & self.mark_bit != 0 {
+                Err(GetError::Closed)
+            } else {
+                Err(GetError::Empty)
+            }
+        }
+        else {
+            // Deconstruct the head.
+            let index = head & (self.mark_bit - 1);
+
+            // Inspect the corresponding slot.
+            let slot = &self.buffer[index];
+            let value = slot.value.with(|slot| unsafe { slot.read().assume_init_read() });
```

Comment:
This code is unsound. This creates a bytewise copy of the head without checking if `T` is `Copy`, which can cause undefined behavior if `T` is `drop`, since the `drop` will be run twice. If we were sending `Box<[u8]>`'s over the `ConcurrentQueue`, this would cause `free()` to be called twice, leading to a double free.

You should instead be using `as_mut_ptr()` and then using `&mut` to cast it to an `&mut T`.

(In addition, since [`assume_init_read`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.assume_init_read) was introduced in 1.60, after the release of 1.48, we can't use it yet).

Reply:
> This code is unsound. This creates a bytewise copy of the head without checking if T is Copy, which can cause undefined behavior if T is drop, since the drop will be run twice. If we were sending Box<[u8]>'s over the ConcurrentQueue, this would cause free() to be called twice, leading to a double free.

Hmm, that makes sense, thanks for the explanation! I haven't done much `unsafe` so far, hence the error.

So in this case we'd have something more or less like:
```
            let slot = &self.buffer[index];
            let value = slot.value.with_mut(|slot| unsafe {
                let ptr = &mut *slot;
                unsafe { &mut *ptr.as_mut_ptr() }
            });
```

And then change the `fn` return type to `Result<&mut T, GetError>`, correct?